agent/eventhandler/attachment_handler.go

```diff
@@ -0,0 +1,158 @@
+	attachmentARNToHandler map[string]*attachmentHandler
+
+	// lock is used to safely access the attachmentARNToHandler map
+	lock sync.Mutex
+
+	client api.ECSClient
+	ctx    context.Context
+}
+
+// attachmentHandler is responsible for handling a certain attachment
+type attachmentHandler struct {
+	// attachmentARN is the arn of the attachment that the attachmentHandler is handling
+	attachmentARN string
+
+	// stateSaver is a statemanager which may be used to save any
+	// changes to an attachment's SentStatus
+	stateSaver statemanager.Saver
+
+	// backoff is the backoff object used in submitting attachment state change
+	backoff retry.Backoff
+
+	// lock is used to ensure that the attached status of an attachment won't be sent multiple times
+	lock sync.Mutex
+
+	client api.ECSClient
+	ctx    context.Context
+}
+
+// NewAttachmentEventHandler returns a new AttachmentEventHandler object
+func NewAttachmentEventHandler(ctx context.Context,
+	stateSaver statemanager.Saver,
+	client api.ECSClient) *AttachmentEventHandler {
+	return &AttachmentEventHandler{
+		ctx:                    ctx,
+		stateSaver:             stateSaver,
+		client:                 client,
+		attachmentARNToHandler: make(map[string]*attachmentHandler),
+		backoff: retry.NewExponentialBackoff(submitStateBackoffMin, submitStateBackoffMax,
+			submitStateBackoffJitterMultiple, submitStateBackoffMultiple),
+	}
+}
+
+// AddStateChangeEvent adds a state change event to AttachmentEventHandler for it to handle
+func (handler *AttachmentEventHandler) AddStateChangeEvent(change statechange.Event) error {
+	if change.GetEventType() != statechange.AttachmentEvent {
+		return fmt.Errorf("eventhandler: attachment handler received unrecognized event type: %d", change.GetEventType())
+	}
+	event, ok := change.(api.AttachmentStateChange)
+	if !ok {
+		return errors.New("eventhandler: unable to get attachment event from state change event")
+	}
+
+	if event.Attachment == nil {
+		return fmt.Errorf("eventhandler: received malformed attachment state change event: %v", event)
+	}
+
+	attachmentARN := event.Attachment.AttachmentARN
+	handler.lock.Lock()
+	if _, ok := handler.attachmentARNToHandler[attachmentARN]; !ok {
+		handler.attachmentARNToHandler[attachmentARN] = &attachmentHandler{
+			attachmentARN: attachmentARN,
+			stateSaver:    handler.stateSaver,
+			client:        handler.client,
+			ctx:           handler.ctx,
+			backoff:       handler.backoff,
+		}
+	}
+	handler.lock.Unlock()
+
+	attachmentHandler := handler.attachmentARNToHandler[attachmentARN]
+	go attachmentHandler.submitAttachmentEvent(&event)
+
+	return nil
+}
+
+// submitAttachmentEvent submits an attachment event to backend
+func (handler *attachmentHandler) submitAttachmentEvent(attachmentChange *api.AttachmentStateChange) {
+	seelog.Debugf("AttachmentHandler: acquiring attachment lock before sending attachment state change for attachment %s", handler.attachmentARN)
+	handler.lock.Lock()
```

Comment:
Just for my understanding, why should this lock be acquired here? It'd be awesome if you could add a comment regarding that as well.

Reply:
i think the udev watcher could send multiple state change events for a certain attachment, for example, one in reconciliation loop and one from udev event. in that case the lock for the attachment handler can prevent sending attachment state change for an attachment multiple times. added comment for this.

Reply:
I see. The `attachmentARNToHandler ` hashmap helps us with deduping that, yeah?

Reply:
yes, with that map, events among different attachments are handled concurrently while events for same attachment is handled serially (therefore won't be sent multiple times)